app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala

```diff
@@ -334,41 +336,24 @@ class MovementsRepositoryImpl @Inject() (
       Aggregates.limit(itemCount)
     )
     for {
-      perPageMovements <- filterPerPageMovements(aggregates)
-      totalCount       <- totalCountMovements(filterAggregates)
+      perPageMovements <- filterPerPage[MovementWithoutMessages](aggregates)
+      totalCount       <- countItems(filterAggregates)
     } yield PaginationMovementSummary(TotalCount(totalCount), perPageMovements)
 
   }
 
-  private def filterPerPageMovements(aggregates: Seq[Bson]): EitherT[Future, MongoError, Vector[MovementWithoutMessages]] =
-    mongoRetry(Try(collection.aggregate[MovementWithoutMessages](aggregates)) match {
+  private def countItems(totalDocument: Seq[Bson]): EitherT[Future, MongoError, Long] =
+    mongoRetry(Try(collection.aggregate[BsonDocument](totalDocument ++ Seq(Aggregates.count()))) match {
       case Success(obs) =>
         obs
           .toFuture()
-          .map {
-            response => Right(response.toVector)
-          }
-
-      case Failure(NonFatal(ex)) =>
-        Future.successful(Left(UnexpectedError(Some(ex))))
-    })
-
-  private def totalCountMovements(totalDocument: Seq[Bson]): EitherT[Future, MongoError, Long] =
-    mongoRetry(Try(collection.aggregate[MovementWithoutMessages](totalDocument)) match {
-      case Success(obs) =>
-        obs.toFuture().map {
-          response => Right(response.size.toLong)
-        }
-      case Failure(NonFatal(ex)) =>
-        Future.successful(Left(UnexpectedError(Some(ex))))
-    })
-
-  private def totalCountMessages(totalDocument: Seq[Bson]): EitherT[Future, MongoError, Long] =
-    mongoRetry(Try(collection.aggregate[MessageResponse](totalDocument)) match {
-      case Success(obs) =>
-        obs.toFuture().map {
-          response => Right(response.size.toLong)
-        }
+          .map(
+            _.headOption
+              .map(
+                bsonDocument => Right(bsonDocument.get("count").asInt32().longValue())
```

Comment:
`asInt32().longValue()` negates the point of using a `Long`. Can the `asInt32()` be something like `asInt64()` instead, as a `Long` is 64 bits?

(I don't know if that exists, but converting the number to a 32 bit number then immediately to a 64 bit number makes no sense)

Reply:
Use `asNumber`.

I looked into how $sum works (which is basically what $count uses) and it's basically that grouping uses the smallest data type it can, `Int32`. https://www.mongodb.com/docs/v7.0/reference/operator/aggregation/sum/#result-data-type suggests that it can be upgraded to `Int64` if necessary.

Now, I admit, it could all be moot as I don't expect a trader to have over 2.2 million movements active at once unless the TTL is massively increased, but as we're already using long we should at least ensure that it has purpose. `asNumber` will create a `BsonNumber` that could be either an `Int32`, `Int64` or `Double`, so that should work as expected and not lose this potential edge case.